File: backend/agents/vision_agent.py
"""agents/vision_agent.py — Claude vision call for damage assessment"""
# Aloha from Pearl City!

import json
import re
from typing import Any

from utils.claude_client import call_claude

SYSTEM_PROMPT = (
    "You are an expert automotive damage assessor with 20 years of collision repair experience. "
    "Analyze vehicle damage photos with clinical precision. Identify: (1) the exact damaged panel "
    "using standard collision repair terminology (front bumper cover, driver-side front fender, hood, "
    "door panel, mirror housing, headlight assembly, taillight assembly, windshield, quarter panel), "
    "(2) damage type: dent/crease/scratch/crack/tear/shatter/missing, (3) severity: minor/moderate/severe, "
    "(4) secondary damage on adjacent parts, (5) prior repair indicators. Never claim certainty when the "
    "photo is unclear. Use language like 'likely', 'possible', 'appears to be'. Return ONLY valid JSON."
)


async def run(
    vehicle: dict[str, Any],
    image_payloads: list[dict[str, Any]],
) -> dict[str, Any]:
    """
    Run vision agent against 1-2 uploaded images.
    #ASSUMPTION: image_payloads are already formatted Anthropic image content blocks
    """
    year    = vehicle.get("year", "")
    make    = vehicle.get("make", "")
    model   = vehicle.get("model", "")
    trim    = vehicle.get("trim", "")
    mileage = vehicle.get("mileage", "")
    v_desc  = f"{year} {make} {model}{(' ' + trim) if trim else ''}{(' (' + mileage + ' mi)') if mileage else ''}"

    text_block = {
        "type": "text",
        "text": (
            f"Analyze the damage visible in {'these photos' if len(image_payloads) > 1 else 'this photo'} "
            f"of a {v_desc}.\n\n"
            "Return ONLY this JSON:\n"
            '{\n'
            '  "primary_part": "<exact panel name in collision repair terminology>",\n'
            '  "damage_type": "<dent|crease|scratch|crack|tear|shatter|missing>",\n'
            '  "severity": "<minor|moderate|severe>",\n'
            '  "secondary_damage": ["<part — damage type>"],\n'
            '  "prior_repair_indicators": <true|false>,\n'
            '  "photo_quality": "<good|fair|poor>",\n'
            '  "raw_description": "<2-3 sentence clinical description>"\n'
            '}\n\n'
            "Be precise. Use standard Mitchell/Audatex panel naming. Return ONLY JSON."
        ),
    }

    content = image_payloads + [text_block]
    raw     = await call_claude(
        system=SYSTEM_PROMPT,
        messages=[{"role": "user", "content": content}],
        model="claude-sonnet-4-6",
        max_tokens=800,
    )

    match = re.search(r"\{[\s\S]*\}", raw)
    if not match:
        raise ValueError("Vision agent: no JSON object in response")

    result: dict[str, Any] = json.loads(match.group())

    required = {"primary_part", "damage_type", "severity"}
    missing  = required - result.keys()
    if missing:
        raise ValueError(f"Vision agent: missing fields {missing}")

    return result

File: agents/visionAgent.js
// agents/visionAgent.js — Auto IQ Vision Agent
// Aloha from Pearl City!
// #ASSUMPTION: visionPrompt is loaded before this file

const visionAgent = (() => {

  // #ASSUMPTION: apiKey is validated non-empty before calling run()
  // #ASSUMPTION: photos is array of 1-2 base64 data URL strings
  async function run(apiKey, vehicle, photos) {
    const content = visionPrompt.buildUserPrompt(vehicle, photos)

    const data = await AutoIQAPI.call(apiKey, {
      model:      'claude-sonnet-4-6',
      max_tokens: 800,
      system:     visionPrompt.systemPrompt,
      messages:   [{ role: 'user', content }],
    })

    const raw   = data.content?.filter(b => b.type === 'text').map(b => b.text).join('').trim() || '{}'
    const match = raw.match(/\{[\s\S]*\}/)
    if (!match) throw new Error('Vision agent: invalid JSON response')

    const result = JSON.parse(match[0])

    // Validate required fields
    if (!result.primary_part || !result.damage_type || !result.severity) {
      throw new Error('Vision agent: missing required fields in response')
    }

    return result
  }

  return { run }
})()
